easybuild/tools/singularity/utilities.py

```diff
@@ -0,0 +1,395 @@
+	
+        environment_content += "module use " +  modulepath + "\n"
+        environment_content +=  "module load " + os.path.join(appname,appver+appversuffix) + "\n"
+
+    if ec_repo:
+    	post_content += "rm -rf easybuild-easyconfigs \n"
+
+    if eb_repo:
+    	post_content += "rm -rf easybuild-easyblocks \n"
+
+    # cleaning up directories in container after build	
+    post_content += """exit
+rm -rf /scratch/tmp/*
+rm -rf /scratch/build
+rm -rf /scratch/sources
+rm -rf /scratch/ebfiles_repo
+"""
+
+
+    runscript_content = """
+%runscript
+eval "$@"
+"""
+
+    label_content = "\n%labels \n"
+
+# uncomment section below to add architecture details in %labels 
+#    label_content += "Architecture " + arch_name + "\n"
+#    label_content += "Host " + system_info['hostname'] + "\n"
+#    label_content += "CPU  " + system_info['cpu_model'] + "\n"
+
+    # adding all the regions for writing the  Singularity definition file
+    content = bootstrap_content + post_content + runscript_content + environment_content + label_content
+    change_dir(singularity_writepath)
+    write_file(def_file,content)
+
+    print "Writing Singularity Definition File: %s" % os.path.join(singularity_writepath,def_file)
+
+    # if easybuild will build container
+    if build_image:
+    	container_name = ""
+
+	if image_name != None:
+		ext =  os.path.splitext(image_name)[1] 
+		if ext == ".img" or ext == ".simg":
+			_log.debug("Extension for image is okay from --image-name")
+		else:
+			print "Invalid Extension for --imagename ", ext
+			EasyBuildError("Invalid Extension for --imagename %s", ext)
+			sys.exit(1)
+
+		container_name = image_name
+	else:
+		# definition file Singularity.<app>-<version, container name <app>-<version>.<img|simg>
+		pos = def_file.find('.')
+		container_name = def_file[pos+1:]
+
+	#squash image format
+	if image_format == "squashfs":
+		container_name += ".simg"
+		if os.path.exists(container_name):
+			errmsg = "Image already exist at " + os.path.join(singularity_writepath,container_name) 
+			print errmsg
+			EasyBuildError(errmsg)
+			sys.exit(0)
+
+		os.system("sudo singularity build " + container_name + " " + def_file)
+
+	# ext3 image format, creating as writable container 
+	elif image_format == "ext3":
+	    	container_name += ".img"
+
+		if os.path.exists(container_name):
+			errmsg = "Image already exist at " + os.path.join(singularity_writepath,container_name) 
+			print errmsg
+			EasyBuildError(errmsg)
+			sys.exit(0)
+
+		os.system("sudo singularity build --writable " + container_name + " " + def_file)
+
+	# sandbox image format, creates as a directory but acts like a container
+	elif image_format == "sandbox":
+
+		if os.path.exists(container_name):
+			errmsg = "Image already exist at " + os.path.join(singularity_writepath,container_name) 
+			print errmsg
+			EasyBuildError(errmsg)
+			sys.exit(0)
+
+	     	os.system("sudo singularity build --sandbox " + container_name + " " + def_file)
```

Comment:
Why mixing using `subprocess` and `os.system` for shell invocation?

Reply:
using subprocess.popen to get return value of singularity version. Using os.system when running shell commands. I was not sure what is the best way to deal with return values and executing shell commands inside python